descholar-ceo/projects-opinions
# Add projects opinions logic and design

## What does this PR do?
It adds all the logic and design needed to implement [Projects Opinions](https://projects-opinions.herokuapp.com/) as its requirements are listed [here](https://www.notion.so/Twitter-redesign-f8a8d48453d54d1a949bb0ceab4c8718)

## Tasks to be completed in this PR
- [x] Project setup
- [x] Add signup and login features
- [x] Add `following` and `unfollowing` feature
- [x] Add post opinions and commenting
- [x] Write test and increase code coverage to at least 95%
- [x] Polish codes and fixing stylelint / Rubocop lint errors

## Deployment
I have deployed this project on heroku [projects opinion](https://projects-opinions.herokuapp.com/)

## How is it tested locally?

### Prerequisites
- [Ruby 2.7.1](https://www.ruby-lang.org/en/news/2020/03/31/ruby-2-7-1-released/)
- [Node v12.18.4 or higher](https://nodejs.org/en/)
- [Yarn v1.22.4 or higher](https://yarnpkg.com/)
- [Git 2.17.0 or higer](https://yarnpkg.com/)
- [Postgresql 9.5.23 or higer](https://www.postgresql.org/)
- Your computer clock should be well set

   ### Steps
- Run `git clone https://github.com/descholar-ceo/projects-opinions && cd projects-opinions` to get a copy of source codes on your local computer and to navigate inside the project directory
- Run `git checkout add-projects-opinions-logic` to checkout this branch
- Run `bundle install && yarn` to install gems and dependencies
- Run `bundle exec rake db:create` to create databases that are being used with this project
- Run `bundle exec rake db:migrate` to migrate (create) all of the tables needed by this project
- Run `bin/rails server` to start the server, normally after the server started, you should be able to access the projects-opinion app on `http://localhost:3000` or `http://127.0.0.1:3000` or `http://0.0.0.0:3000` in your browser

__NOTE:__ If you don't want to go through all of the above steps, you can just access this project online on [this link](https://projects-opinions.herokuapp.com/)

### Tests
- Create a file in `config/` folder and call it `master.key`
- You will need a master key to paste in that file
- Make sure that your computer clock is on the right time
- Run this command `bundle exec rspec` to run tests

## Added feature
I have added some features which were not listed amongst the project requirements, those are:
- User can comment on opinions
- User can unfollow user that they follow
- User can signing in properly, (Authentication using [devise](https://github.com/heartcombo/devise))
